<?xml version="1.0"?>
<launch>
    
    <arg name="robot_name"            default="$(env ROBOT_NAME)" />
    <arg name="base_frame"            default="base_link" />
    <arg name="output"                default="screen" />
    <arg name="gui_publish_rate"      default="-1.0" />
    <arg name="M_PI"                  value="3.141592654"/>

    <node pkg="amcl" type="amcl" name="amcl" output="screen" respawn="true">
        <!-- Publish scans from best pose at a max of 10 Hz -->
        <param name="odom_frame_id" value="$(arg robot_name)/odom"/>
        <param name="global_frame_id" value="$(arg robot_name)/map"/>
        <param name="base_frame_id" value="$(arg robot_name)/$(arg base_frame)"/>
        
        <param name="use_map_topic" value="true"/>
        <param name="first_map_only" value="true"/>
        <param name="tf_broadcast" value="true"/>
    
        <param name="initial_cov_xx" value="0.24278913053989015" />
        <param name="initial_cov_yy" value="0.24259989395034498" />
        <param name="initial_cov_aa" value="0.0650658130276735" />
        <param name="initial_pose_x" value="-8.065282749033946" />
        <param name="initial_pose_y" value="0.37984209874683633" />
        <param name="initial_pose_a" value="-0.6427092977146723" />

        <!-- <param name="initial_cov_xx" value="0.0018985060089144383" />
        <param name="initial_cov_yy" value="0.001923474488618293" />
        <param name="initial_cov_aa" value="0.0015270140188449867" />
        <param name="initial_pose_x" value="0.045338025084146416" />
        <param name="initial_pose_y" value="1.0412823626320105" />
        <param name="initial_pose_a" value="0.01953112627741272" /> -->

        <!-- <param name="initial_cov_xx" value="0.0018415420343025346" />
        <param name="initial_cov_yy" value="0.0022896819177560257" />
        <param name="initial_cov_aa" value="0.0028981797485869974" />
        <param name="initial_pose_x" value="-5.061254663547058" />
        <param name="initial_pose_y" value="-0.6153396339888759" />
        <param name="initial_pose_a" value="-0.03228037485998501" />
         -->
        <!-- <param name="initial_cov_xx" value="0.003331837481003852" />
        <param name="initial_cov_yy" value="0.003267908361100247" />
        <param name="initial_cov_aa" value="0.003647550699637466" />
        <param name="initial_pose_x" value="1.5039816812104185" />
        <param name="initial_pose_y" value="2.17038626221865" />
        <param name="initial_pose_a" value="-3.085284075384399" /> -->

        <param name="gui_publish_rate" value="$(arg gui_publish_rate)"/>

        <param name="min_particles" value="100"/>
        <param name="max_particles" value="5000"/>
        <param name="kld_err" value="0.10"/> <!-- 0.01 -->
        <param name="kld_z" value="0.5"/> <!-- 0.99 -->
        <param name="update_min_d" value="0.1"/>
        <param name="update_min_a" value="$(eval 5.0 * arg('M_PI') / 180.0)"/>
        <!-- <param name="update_min_d" value="0.10"/>
        <param name="update_min_a" value="$(eval 5.0 * arg('M_PI') / 180.0)"/> -->
        <param name="resample_interval" value="2"/>
        <param name="transform_tolerance" value="1.0" />
        <param name="recovery_alpha_slow" value="0.0"/> <!-- 0.001 -->
        <param name="recovery_alpha_fast" value="0.0"/> <!-- 0.1 -->
        <param name="selective_resampling" value="false"/> <!-- true -->
        <param name="save_pose_rate" value="1.0"/>
            
        <param name="odom_model_type" value="diff-corrected"/>
        <param name="odom_alpha1" value="0.008"/> <!-- 0.005 -->
        <param name="odom_alpha2" value="0.040"/> <!-- 0.005 -->
        <param name="odom_alpha3" value="0.004"/> <!-- 0.005 -->
        <param name="odom_alpha4" value="0.025"/> <!-- 0.005 -->
        <param name="odom_alpha5" value="0.005"/>

        <param name="laser_min_range" value="-1.0"/>
        <param name="laser_max_range" value="-1.0"/>
        <param name="laser_max_beams"  value="250"/> <!-- 250 -->
        <param name="laser_z_hit" value="0.9"/> <!-- 0.95 -->
        <param name="laser_z_short" value="0.1"/>
        <param name="laser_z_max" value="0.05"/>
        <param name="laser_z_rand" value="0.05"/> <!-- 0.05 -->
        <param name="laser_sigma_hit" value="0.1"/> <!-- 0.2 -->
        <param name="laser_lambda_short" value="0.1"/>
        <param name="laser_model_type" value="likelihood_field"/>
        <param name="laser_likelihood_max_dist" value="4.0"/> <!-- 4.0 -->
        <!-- <rosparam file="$(find leo_robot)/config/navigation/amcl_initial_pose.yaml" command="load"/> -->
    </node>

</launch>
